openedx/enterprise-access
# feat: introduce earliest_expiry_date on assignments, expired assignment state, API to acknowledge expired/cancelled assignments, and more

# Description

Ticket: https://2u-internal.atlassian.net/browse/ENT-8162

Related ADR: https://github.com/openedx/enterprise-access/blob/main/docs/decisions/0015-expiration-improvements.rst

Related PR: https://github.com/openedx/frontend-app-learner-portal-enterprise/pull/932

## Changelog

* Introduces new `expired` state for `LearnerContentAssigment` records.
* Introduces the following model fields on `LearnerContentAssignment` records:
  * `allocated_at`
  * `accepted_at`
  * `errored_at`
  * `cancelled_at`
  * `expired_at`
* Removes unused `last_notification_at` field.
* Replaces new `automatic_cancellation` assignment actions with an `expired` assignment actions instead.
* Introduces two new actions related to assignment acknowledgement/dismissal:
  * `cancelled_acknowledged`: "Learner assignment cancellation acknowledged by learner"
  * `expired_acknowledged`: "Learner assignment expiration acknowledged by learner"
* Implements an `acknowledge-assignments` POST API endpoint, that accepts a list of `assignment_uuids` to acknowledge. An appropriate action is created depending on whether the assignment `state` is `expired` or `cancelled`.
  * Utilizes new permission specific to assignment acknowledgement, available to learners/admins/operators for assignments they are linked to through their `lms_user_id`.
* Writes a data migration to migrate "automatic cancellation" actions to "expired" actions and change the associated assignment record's state to `expired` (if currently marked as cancelled).
* Updates the job that automatically expires assignments to set the expired assignments' states to `expired`.
* Introduces a new `learner_state="expired"` field for assignments in an `expired` state.
* Serializes `earliest_possible_expiration` on assignment records (impacts both `credits_available` and the `admin/assignments/` list endpoints).
* Refactors to consolidate business logic around triple-date expiration into one method, usable in multiple existing places where duplicate logic lived. Introduced the consolidated logic to avoid introducing a third duplicate place.

```json
{
    "earliest_possible_expiration": {
        "date": "2023-06-22T00:00:00Z",
        "reason": "SUBSIDY_EXPIRED"
    }
}
```

* Serializes boolean `learner_acknowledged` on assignments and associated assignment `actions`.
* Ensures already acknowledged assignments are NOT returned in the `credits_available` API, such that they no longer appear for learners once acknowledged/dismissed.
* Expired assignments will NOT appear in admin portal Learner Credit Management as the "Assigned" table only looks at `state="allocated"` or `state="errored"`.

## Examples and screenshots

New POST API endpoint `/api/v1/assignment-configurations/<uuid>/acknowledge-assignments/`, accepting the following JSON payload:

```json
{
    "assignment_uuids": ["38c4269e-f902-4066-b628-860493613138"]
}
```

Screenshot of an assignment that has been expired, and then acknowledged through the API. 

<img width="505" alt="image" src="https://github.com/openedx/enterprise-access/assets/2828721/dced2616-0744-4f88-9c13-a99f7ecdfe53">

Screenshot of an assignment that has been cancelled, and then acknowledged through the API.

<img width="505" alt="image" src="https://github.com/openedx/enterprise-access/assets/2828721/e25c3e71-7d37-442d-a336-12fb84ebebd3">

Example API response where provided assignment UUID was successfully acknowledged:

```json
{
    "acknowledged_assignments": [
        "38c4269e-f902-4066-b628-860493613138"
    ],
    "already_acknowledged_assignments": [],
    "unacknowledged_assignments": []
}
```

Example API response where provided assignment UUID has already been acknowledged previously:

```json
{
    "acknowledged_assignments": [],
    "already_acknowledged_assignments": [
        "38c4269e-f902-4066-b628-860493613138"
    ],
    "unacknowledged_assignments": []
}
```

## Testing instructions

Note: the below instructions expire assignment(s) with expired subsidy; you may also expire assignment via 90 days rule or past content enrollment deadline as well.

1. Provision (or re-use) an assignable Policy that is associated with an non-expired Subsidy.
1. Assign assignment(s) to at least 1 user.
1. Edit the Subsidy to be expired (expiration date in the past).
1. Run the `./manage.py automatically_expire_assignments`.
1. Ensure the associated [Learner Portal PR](https://github.com/openedx/frontend-app-learner-portal-enterprise/pull/932) is running.
1. Observe now-expired assignment on Learner Portal dashboard page route with related `Alert` messaging.
1. Dismissing the  related `Alert` messaging should make API call to `acknowledge-assignments` for the expired assignment, and then make another call to `credits_available` to refresh the query cache, thus removing the now-acknowledged assignments from the dashboard UI.
